packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -309,267 +333,176 @@ class FadeInImage extends StatefulWidget {
-
-  /// Fading in new image.
-  fadeIn,
-
-  /// Fade-in complete.
-  completed,
-}
-
-typedef _ImageProviderResolverListener = void Function();
-
-class _ImageProviderResolver {
-  _ImageProviderResolver({
-    @required this.state,
-    @required this.listener,
-  });
-
-  final _FadeInImageState state;
-  final _ImageProviderResolverListener listener;
-
-  FadeInImage get widget => state.widget;
-
-  ImageStream _imageStream;
-  ImageInfo _imageInfo;
-
-  void resolve(ImageProvider provider) {
-    final ImageStream oldImageStream = _imageStream;
-    _imageStream = provider.resolve(createLocalImageConfiguration(
-      state.context,
-      size: widget.width != null && widget.height != null ? Size(widget.width, widget.height) : null,
-    ));
-    assert(_imageStream != null);
-
-    if (_imageStream.key != oldImageStream?.key) {
-      final ImageStreamListener listener = ImageStreamListener(_handleImageChanged);
-      oldImageStream?.removeListener(listener);
-      _imageStream.addListener(listener);
-    }
-  }
-
-  void _handleImageChanged(ImageInfo imageInfo, bool synchronousCall) {
-    _imageInfo = imageInfo;
-    listener();
-  }
-
-  void stopListening() {
-    _imageStream?.removeListener(ImageStreamListener(_handleImageChanged));
-  }
-}
-
 class _FadeInImageState extends State<FadeInImage> with TickerProviderStateMixin {
-  _ImageProviderResolver _imageResolver;
-  _ImageProviderResolver _placeholderResolver;
-
   AnimationController _controller;
-  Animation<double> _animation;
-
-  FadeInImagePhase _phase = FadeInImagePhase.start;
-  FadeInImagePhase get phase => _phase;
+  Animation<double> _fadeOutAnimation;
+  Animation<double> _fadeInAnimation;
 
   @override
   void initState() {
-    _imageResolver = _ImageProviderResolver(state: this, listener: _updatePhase);
-    _placeholderResolver = _ImageProviderResolver(state: this, listener: () {
-      setState(() {
-        // Trigger rebuild to display the placeholder image
-      });
-    });
-    _controller = AnimationController(
-      value: 1.0,
-      vsync: this,
-    );
-    _controller.addListener(() {
-      setState(() {
-        // Trigger rebuild to update opacity value.
-      });
-    });
-    _controller.addStatusListener((AnimationStatus status) {
-      _updatePhase();
-    });
     super.initState();
+    _controller = AnimationController(vsync: this);
+    _initializeAnimations();
   }
 
   @override
-  void didChangeDependencies() {
-    _resolveImage();
-    super.didChangeDependencies();
+  void dispose() {
+    _fadeInAnimation.removeStatusListener(_handleFadeInAnimationStatusChange);
+    _fadeInAnimation.removeListener(_handleFadeInTick);
+    _controller.dispose();
+    super.dispose();
   }
 
   @override
   void didUpdateWidget(FadeInImage oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (widget.image != oldWidget.image || widget.placeholder != oldWidget.placeholder)
-      _resolveImage();
+    if (!_controller.isAnimating &&
```

Comment:
why only if it's not animating?

Reply:
I thought it would yield a jarring fade if the animation values changed mid-animation.  But I realize that I still need to update them either way, which I'm failing todo here.

Thoughts?  Better to update mid-animation and make the code simple, or better to wait until the existing animation is done, at the cost of complexity?

Reply:
Not relevant anymore.